subsys/bluetooth/host/hci_core.c

```diff
@@ -3370,6 +3374,42 @@ static void le_adv_report(struct net_buf *buf)
 }
 #endif /* CONFIG_BT_OBSERVER */
 
+#if defined(CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK)
+int bt_hci_register_vnd_evt_cb(bt_hci_vnd_evt_cb_t cb)
+{
+	hci_vnd_evt_cb = cb;
+	return 0;
+}
+#endif /* CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK */
+
+static inline void default_hci_vendor_event(struct net_buf *buf)
+{
+	/* do nothing at present time */
+	BT_WARN("Unhandled vendor-specific event: %s",
+		bt_hex(buf->data, buf->len));
+}
+
+static void hci_vendor_event(struct net_buf *buf)
+{
+	int err = -EINVAL;
+
+#if defined(CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK)
+	if (hci_vnd_evt_cb) {
+		struct net_buf_simple_state state;
+
+		net_buf_simple_save(&buf->b, &state);
+
+		err = hci_vnd_evt_cb(&buf->b);
+
+		net_buf_simple_restore(&buf->b, &state);
+	}
+#endif /* CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK */
+
+	if (err) {
```

Comment:
See  my earlier comment, I would probably do:
```
if (err == -ENOTSUP) {
   default_hci_vendor_event()
}
else {
  return err;
}

Reply:
There is nothing to return an error code back to; you cannot indicate to the controller that the event wasn't handled.

Reply:
OK, good point. Then I'm happy if we replace EINVAL with ENOTSUP and document it properly

Reply:
@carlescufi What about other return values then?

```
0 : hook handled the event
-ENOTSUP : defer the handling of this vendor-specific event to Zephyr's Bluetooth stack
any other value : ???? (maybe assert because the hook is not following the defined contract)
```

Or maybe we should just say that 0 mean that the hook has handled the value and any value other that 0 means "defer the handling of this vendor-specific event to Zephyr's Bluetooth stack"

Reply:
Or, just define an `enum { BT_HCI_VND_EVT_HANDLED, BT_HCI_VND_EVT_NOT_HANDLED }` and have the function return that?

Reply:
OK, in that case why don't we just use a bool as a return value?

Reply:
I used int because all the other functions in hci.h returns int and have "@return 0 on success or negative error value on failure.", but if bool is okay as a return value then i'm in favor of that.

Reply:
> in that case why don't we just use a bool as a return value?

Readibility, I suppose. I generally try to use bool only when the name of the function is inherently boolean, i.e. it's clear from the name what true/false means. I'm not sure that's true here. That said, it's of course some overhead to add a new enum definition, so I don't have a super strong opinion on this.

Reply:
@thoh-ot in this case we're not really dealing with a "failure or succcess" case, which is what int is normally used for. Rather, both possible return values can mean that everything went as expected.

Reply:
What about just saying:

0 : hook handled the vendor-specific event
any other value: defer the handling of this vendor-specific event to Zephyr's Bluetooth stack